src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,196 @@
-// Saddleback search is a technique used to find an element in a sorted 2D matrix in O(m + n) time,
-// where m is the number of rows, and n is the number of columns. It works by starting from the
-// top-right corner of the matrix and moving left or down based on the comparison of the current
-// element with the target element.
 use std::cmp::Ordering;
 
-pub fn saddleback_search(matrix: &[Vec<i32>], element: i32) -> (usize, usize) {
-    // Initialize left and right indices
+/// Custom error type to represent errors related to matrix validation.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixError {
+    NonRectangularMatrix,
+}
+
+/// Performs Saddleback search on a sorted 2D matrix.
```

Comment:
What do you think about introducing a `is_sorted` function, checking if the matrix is sorted as expected. Clearly if it would be used inside the `saddleback_search` function it would spoil the complexity, but if would be great if the user could verify if their matrix is as expected. I guess the possible workflow could look like this:
- the used generates some data,
- check the data with `is_sorted`,
- uses many times the `saddleback_search`.

Having such function would also _define_ what does it mean to have a _sorted matrix_.

Reply:
Adding an `is_sorted` function is a great idea. It allows users to validate whether their matrix is correctly sorted before running the `saddleback_search` and provides a clear definition of a _sorted matrix_. It can be useful when users generate their data or when matrices are passed from external sources to ensure the structure is valid.